File: swift/Sources/Shield/IdentityProvider.swift
import Foundation
import CommonCrypto

/// User identity.
public struct Identity {
    public let userId: String
    public let displayName: String
    public let verificationKey: [UInt8]
    public let createdAt: Int64
    public var attributes: [String: Any]

    public init(userId: String, displayName: String, verificationKey: [UInt8],
                createdAt: Int64, attributes: [String: Any] = [:]) {
        self.userId = userId
        self.displayName = displayName
        self.verificationKey = verificationKey
        self.createdAt = createdAt
        self.attributes = attributes
    }
}

/// Session information from validated token.
public struct Session {
    public let userId: String
    public let created: Int64
    public let expires: Int64
    public let permissions: [String]
    public var metadata: [String: Any]

    public init(userId: String, created: Int64, expires: Int64,
                permissions: [String] = [], metadata: [String: Any] = [:]) {
        self.userId = userId
        self.created = created
        self.expires = expires
        self.permissions = permissions
        self.metadata = metadata
    }

    public var isExpired: Bool {
        return Int64(Date().timeIntervalSince1970) > expires
    }

    public var remainingTime: Int64 {
        return max(0, expires - Int64(Date().timeIntervalSince1970))
    }
}

/// IdentityProvider - SSO/Identity Provider using symmetric crypto.
///
/// Provides user registration, session management, and service tokens
/// using only symmetric cryptography (no public-key certificates).
public class IdentityProvider {
    private static let pbkdf2Iterations = 100000

    private let providerKey: [UInt8]
    private let tokenTtl: Int
    private var identities: [String: Identity] = [:]

    /// Create identity provider.
    ///
    /// - Parameters:
    ///   - providerKey: 32-byte provider secret key
    ///   - tokenTtl: Default token lifetime in seconds
    public init(providerKey: [UInt8], tokenTtl: Int = 3600) {
        self.providerKey = providerKey
        self.tokenTtl = tokenTtl > 0 ? tokenTtl : 3600
    }

    /// Register new user identity.
    ///
    /// - Parameters:
    ///   - userId: Unique user identifier
    ///   - password: User's password
    ///   - displayName: User's display name
    ///   - attributes: Optional user attributes
    /// - Returns: Created identity
    /// - Throws: ShieldError if userId already exists
    public func register(userId: String, password: String, displayName: String,
                         attributes: [String: Any] = [:]) throws -> Identity {
        if identities[userId] != nil {
            throw ShieldError.invalidToken
        }

        let verificationKey = deriveVerificationKey(userId: userId, password: password)
        let identity = Identity(
            userId: userId,
            displayName: displayName,
            verificationKey: verificationKey,
            createdAt: Int64(Date().timeIntervalSince1970),
            attributes: attributes
        )

        identities[userId] = identity
        return identity
    }

    /// Authenticate user and return session token.
    ///
    /// - Parameters:
    ///   - userId: User identifier
    ///   - password: User's password
    ///   - permissions: Optional permission list
    ///   - ttl: Token lifetime (or default if nil)
    /// - Returns: Session token, or nil if authentication fails
    public func authenticate(userId: String, password: String,
                             permissions: [String]? = nil, ttl: Int? = nil) -> String? {
        guard let identity = identities[userId] else {
            return nil
        }

        let verificationKey = deriveVerificationKey(userId: userId, password: password)
        guard Shield.constantTimeEquals(verificationKey, identity.verificationKey) else {
            return nil
        }

        let actualTtl = ttl ?? tokenTtl
        let now = Int64(Date().timeIntervalSince1970)

        var sessionData: [String: Any] = [
            "user_id": userId,
            "created": now,
            "expires": now + Int64(actualTtl),
            "permissions": permissions ?? [],
            "nonce": generateNonce()
        ]

        return signToken(data: sessionData)
    }

    /// Validate session token.
    ///
    /// - Parameter token: Session token from authenticate()
    /// - Returns: Session object, or nil if invalid/expired
    public func validateToken(_ token: String) -> Session? {
        guard let sessionData = verifyToken(token) else {
            return nil
        }

        guard let expires = sessionData["expires"] as? Int64 else {
            return nil
        }

        if expires < Int64(Date().timeIntervalSince1970) {
            return nil
        }

        let permissions = (sessionData["permissions"] as? [String]) ?? []
        let metadata = (sessionData["metadata"] as? [String: Any]) ?? [:]

        return Session(
            userId: sessionData["user_id"] as? String ?? "",
            created: sessionData["created"] as? Int64 ?? 0,
            expires: expires,
            permissions: permissions,
            metadata: metadata
        )
    }

    /// Create service-specific access token.
    ///
    /// - Parameters:
    ///   - sessionToken: Valid session token
    ///   - service: Target service identifier
    ///   - permissions: Scoped permissions for this service
    ///   - ttl: Token lifetime (default 300 seconds)
    /// - Returns: Service token, or nil if session invalid
    public func createServiceToken(sessionToken: String, service: String,
                                   permissions: [String]? = nil, ttl: Int = 300) -> String? {
        guard let session = validateToken(sessionToken) else {
            return nil
        }

        let now = Int64(Date().timeIntervalSince1970)
        let serviceData: [String: Any] = [
            "user_id": session.userId,
            "service": service,
            "created": now,
            "expires": now + Int64(ttl),
            "permissions": permissions ?? [],
            "parent_expires": session.expires
        ]

        return signToken(data: serviceData)
    }

    /// Validate service-specific token.
    ///
    /// - Parameters:
    ///   - token: Service token
    ///   - service: Expected service identifier
    /// - Returns: Session object, or nil if invalid
    public func validateServiceToken(_ token: String, service: String) -> Session? {
        guard let tokenData = verifyToken(token) else {
            return nil
        }

        guard tokenData["service"] as? String == service else {
            return nil
        }

        let now = Int64(Date().timeIntervalSince1970)
        guard let expires = tokenData["expires"] as? Int64, expires >= now else {
            return nil
        }

        if let parentExpires = tokenData["parent_expires"] as? Int64, parentExpires < now {
            return nil
        }

        let permissions = (tokenData["permissions"] as? [String]) ?? []
        let metadata: [String: Any] = ["service": service]

        return Session(
            userId: tokenData["user_id"] as? String ?? "",
            created: tokenData["created"] as? Int64 ?? 0,
            expires: expires,
            permissions: permissions,
            metadata: metadata
        )
    }

    /// Refresh session token.
    ///
    /// - Parameters:
    ///   - token: Current valid session token
    ///   - ttl: New lifetime (or default if nil)
    /// - Returns: New session token, or nil if current token invalid
    public func refreshToken(_ token: String, ttl: Int? = nil) -> String? {
        guard let session = validateToken(token) else {
            return nil
        }

        let actualTtl = ttl ?? tokenTtl
        let now = Int64(Date().timeIntervalSince1970)

        let sessionData: [String: Any] = [
            "user_id": session.userId,
            "created": now,
            "expires": now + Int64(actualTtl),
            "permissions": session.permissions,
            "nonce": generateNonce()
        ]

        return signToken(data: sessionData)
    }

    /// Revoke user identity.
    ///
    /// - Parameter userId: User to revoke
    /// - Returns: true if user was revoked
    @discardableResult
    public func revokeUser(_ userId: String) -> Bool {
        return identities.removeValue(forKey: userId) != nil
    }

    /// Get identity by user ID.
    public func getIdentity(_ userId: String) -> Identity? {
        return identities[userId]
    }

    // Private helpers

    private func deriveVerificationKey(userId: String, password: String) -> [UInt8] {
        let saltInput = "user:\(userId)".data(using: .utf8)!
        let salt = Shield.sha256(Array(saltInput))

        let userKey = pbkdf2(password: password, salt: salt, iterations: IdentityProvider.pbkdf2Iterations, keyLength: 32)

        var verifyInput = Array("verify:".utf8)
        verifyInput.append(contentsOf: userKey)
        return Shield.sha256(verifyInput)
    }

    private func pbkdf2(password: String, salt: [UInt8], iterations: Int, keyLength: Int) -> [UInt8] {
        var derivedKey = [UInt8](repeating: 0, count: keyLength)
        let passwordData = password.data(using: .utf8)!

        passwordData.withUnsafeBytes { passwordBytes in
            salt.withUnsafeBytes { saltBytes in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passwordBytes.baseAddress?.assumingMemoryBound(to: Int8.self),
                    passwordData.count,
                    saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                    salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(iterations),
                    &derivedKey,
                    keyLength
                )
            }
        }

        return derivedKey
    }

    private func signToken(data: [String: Any]) -> String? {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: data, options: [.sortedKeys]) else {
            return nil
        }

        let tokenBytes = Array(jsonData)
        let mac = Shield.hmacSha256(key: providerKey, data: tokenBytes)
        let truncatedMac = Array(mac.prefix(16))

        var result = tokenBytes
        result.append(contentsOf: truncatedMac)

        return Data(result).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private func verifyToken(_ token: String) -> [String: Any]? {
        // Restore base64 standard chars and padding
        var base64 = token
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }

        guard let decoded = Data(base64Encoded: base64) else {
            return nil
        }

        let bytes = Array(decoded)
        if bytes.count < 17 {
            return nil
        }

        let tokenBytes = Array(bytes.prefix(bytes.count - 16))
        let receivedMac = Array(bytes.suffix(16))

        let expectedMac = Array(Shield.hmacSha256(key: providerKey, data: tokenBytes).prefix(16))

        guard Shield.constantTimeEquals(receivedMac, expectedMac) else {
            return nil
        }

        guard let json = try? JSONSerialization.jsonObject(with: Data(tokenBytes), options: []) as? [String: Any] else {
            return nil
        }

        return json
    }

    private func generateNonce() -> String {
        guard let bytes = Shield.randomBytes(8) else {
            return ""
        }
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}

File: csharp/Shield/IdentityProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Guard8.Shield
{
    /// <summary>
    /// User identity.
    /// </summary>
    public class Identity
    {
        public string UserId { get; }
        public string DisplayName { get; }
        public byte[] VerificationKey { get; }
        public long CreatedAt { get; }
        public Dictionary<string, object> Attributes { get; }

        public Identity(string userId, string displayName, byte[] verificationKey,
                        long createdAt, Dictionary<string, object> attributes = null)
        {
            UserId = userId;
            DisplayName = displayName;
            VerificationKey = verificationKey;
            CreatedAt = createdAt;
            Attributes = attributes ?? new Dictionary<string, object>();
        }
    }

    /// <summary>
    /// Session information from validated token.
    /// </summary>
    public class Session
    {
        public string UserId { get; }
        public long Created { get; }
        public long Expires { get; }
        public List<string> Permissions { get; }
        public Dictionary<string, object> Metadata { get; }

        public Session(string userId, long created, long expires,
                       List<string> permissions = null, Dictionary<string, object> metadata = null)
        {
            UserId = userId;
            Created = created;
            Expires = expires;
            Permissions = permissions ?? new List<string>();
            Metadata = metadata ?? new Dictionary<string, object>();
        }

        public bool IsExpired => DateTimeOffset.UtcNow.ToUnixTimeSeconds() > Expires;

        public long RemainingTime => Math.Max(0, Expires - DateTimeOffset.UtcNow.ToUnixTimeSeconds());
    }

    /// <summary>
    /// IdentityProvider - SSO/Identity Provider using symmetric crypto.
    ///
    /// Provides user registration, session management, and service tokens
    /// using only symmetric cryptography (no public-key certificates).
    /// </summary>
    public class IdentityProvider
    {
        private const int PBKDF2_ITERATIONS = 100000;

        private readonly byte[] _providerKey;
        private readonly int _tokenTtl;
        private readonly Dictionary<string, Identity> _identities = new();

        /// <summary>
        /// Create identity provider.
        /// </summary>
        /// <param name="providerKey">32-byte provider secret key</param>
        /// <param name="tokenTtl">Default token lifetime in seconds</param>
        public IdentityProvider(byte[] providerKey, int tokenTtl = 3600)
        {
            _providerKey = (byte[])providerKey.Clone();
            _tokenTtl = tokenTtl > 0 ? tokenTtl : 3600;
        }

        /// <summary>
        /// Register new user identity.
        /// </summary>
        /// <param name="userId">Unique user identifier</param>
        /// <param name="password">User's password</param>
        /// <param name="displayName">User's display name</param>
        /// <param name="attributes">Optional user attributes</param>
        /// <returns>Created identity</returns>
        public Identity Register(string userId, string password, string displayName,
                                 Dictionary<string, object> attributes = null)
        {
            if (_identities.ContainsKey(userId))
                throw new ArgumentException($"User {userId} already exists");

            byte[] verificationKey = DeriveVerificationKey(userId, password);
            var identity = new Identity(
                userId,
                displayName,
                verificationKey,
                DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                attributes
            );

            _identities[userId] = identity;
            return identity;
        }

        /// <summary>
        /// Authenticate user and return session token.
        /// </summary>
        /// <param name="userId">User identifier</param>
        /// <param name="password">User's password</param>
        /// <param name="permissions">Optional permission list</param>
        /// <param name="ttl">Token lifetime (or default if null)</param>
        /// <returns>Session token, or null if authentication fails</returns>
        public string Authenticate(string userId, string password,
                                   List<string> permissions = null, int? ttl = null)
        {
            if (!_identities.TryGetValue(userId, out var identity))
                return null;

            byte[] verificationKey = DeriveVerificationKey(userId, password);
            if (!Shield.ConstantTimeEquals(verificationKey, identity.VerificationKey, 32))
                return null;

            int actualTtl = ttl ?? _tokenTtl;
            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();

            var sessionData = new Dictionary<string, object>
            {
                ["user_id"] = userId,
                ["created"] = now,
                ["expires"] = now + actualTtl,
                ["permissions"] = permissions ?? new List<string>(),
                ["nonce"] = GenerateNonce()
            };

            return SignToken(sessionData);
        }

        /// <summary>
        /// Validate session token.
        /// </summary>
        /// <param name="token">Session token from Authenticate()</param>
        /// <returns>Session object, or null if invalid/expired</returns>
        public Session ValidateToken(string token)
        {
            var sessionData = VerifyToken(token);
            if (sessionData == null)
                return null;

            long expires = GetLong(sessionData, "expires");
            if (expires < DateTimeOffset.UtcNow.ToUnixTimeSeconds())
                return null;

            var permissions = GetStringList(sessionData, "permissions");
            var metadata = GetDict(sessionData, "metadata");

            return new Session(
                sessionData["user_id"].ToString(),
                GetLong(sessionData, "created"),
                expires,
                permissions,
                metadata
            );
        }

        /// <summary>
        /// Create service-specific access token.
        /// </summary>
        /// <param name="sessionToken">Valid session token</param>
        /// <param name="service">Target service identifier</param>
        /// <param name="permissions">Scoped permissions for this service</param>
        /// <param name="ttl">Token lifetime (default 300 seconds)</param>
        /// <returns>Service token, or null if session invalid</returns>
        public string CreateServiceToken(string sessionToken, string service,
                                         List<string> permissions = null, int ttl = 300)
        {
            var session = ValidateToken(sessionToken);
            if (session == null)
                return null;

            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var serviceData = new Dictionary<string, object>
            {
                ["user_id"] = session.UserId,
                ["service"] = service,
                ["created"] = now,
                ["expires"] = now + ttl,
                ["permissions"] = permissions ?? new List<string>(),
                ["parent_expires"] = session.Expires
            };

            return SignToken(serviceData);
        }

        /// <summary>
        /// Validate service-specific token.
        /// </summary>
        /// <param name="token">Service token</param>
        /// <param name="service">Expected service identifier</param>
        /// <returns>Session object, or null if invalid</returns>
        public Session ValidateServiceToken(string token, string service)
        {
            var tokenData = VerifyToken(token);
            if (tokenData == null)
                return null;

            if (tokenData["service"]?.ToString() != service)
                return null;

            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            long expires = GetLong(tokenData, "expires");
            if (expires < now)
                return null;

            if (tokenData.ContainsKey("parent_expires"))
            {
                long parentExpires = GetLong(tokenData, "parent_expires");
                if (parentExpires < now)
                    return null;
            }

            var permissions = GetStringList(tokenData, "permissions");
            var metadata = new Dictionary<string, object> { ["service"] = service };

            return new Session(
                tokenData["user_id"].ToString(),
                GetLong(tokenData, "created"),
                expires,
                permissions,
                metadata
            );
        }

        /// <summary>
        /// Refresh session token.
        /// </summary>
        /// <param name="token">Current valid session token</param>
        /// <param name="ttl">New lifetime (or default if null)</param>
        /// <returns>New session token, or null if current token invalid</returns>
        public string RefreshToken(string token, int? ttl = null)
        {
            var session = ValidateToken(token);
            if (session == null)
                return null;

            int actualTtl = ttl ?? _tokenTtl;
            long now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();

            var sessionData = new Dictionary<string, object>
            {
                ["user_id"] = session.UserId,
                ["created"] = now,
                ["expires"] = now + actualTtl,
                ["permissions"] = session.Permissions,
                ["nonce"] = GenerateNonce()
            };

            return SignToken(sessionData);
        }

        /// <summary>
        /// Revoke user identity.
        /// </summary>
        /// <param name="userId">User to revoke</param>
        /// <returns>true if user was revoked</returns>
        public bool RevokeUser(string userId)
        {
            return _identities.Remove(userId);
        }

        /// <summary>
        /// Get identity by user ID.
        /// </summary>
        public Identity GetIdentity(string userId)
        {
            return _identities.TryGetValue(userId, out var identity) ? identity : null;
        }

        // Private helpers

        private byte[] DeriveVerificationKey(string userId, string password)
        {
            using var sha256 = SHA256.Create();
            byte[] salt = sha256.ComputeHash(Encoding.UTF8.GetBytes($"user:{userId}"));

            using var pbkdf2 = new Rfc2898DeriveBytes(password, salt, PBKDF2_ITERATIONS, HashAlgorithmName.SHA256);
            byte[] userKey = pbkdf2.GetBytes(32);

            byte[] verifyInput = Encoding.UTF8.GetBytes("verify:").Concat(userKey).ToArray();
            return sha256.ComputeHash(verifyInput);
        }

        private string SignToken(Dictionary<string, object> data)
        {
            string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = false });
            byte[] tokenBytes = Encoding.UTF8.GetBytes(json);

            using var hmac = new HMACSHA256(_providerKey);
            byte[] fullMac = hmac.ComputeHash(tokenBytes);
            byte[] truncatedMac = fullMac.Take(16).ToArray();

            byte[] result = new byte[tokenBytes.Length + 16];
            Array.Copy(tokenBytes, 0, result, 0, tokenBytes.Length);
            Array.Copy(truncatedMac, 0, result, tokenBytes.Length, 16);

            return Convert.ToBase64String(result).Replace('+', '-').Replace('/', '_').TrimEnd('=');
        }

        private Dictionary<string, object> VerifyToken(string token)
        {
            try
            {
                // Restore base64 padding and URL-safe chars
                token = token.Replace('-', '+').Replace('_', '/');
                switch (token.Length % 4)
                {
                    case 2: token += "=="; break;
                    case 3: token += "="; break;
                }

                byte[] decoded = Convert.FromBase64String(token);
                if (decoded.Length < 17)
                    return null;

                byte[] tokenBytes = decoded.Take(decoded.Length - 16).ToArray();
                byte[] receivedMac = decoded.Skip(decoded.Length - 16).ToArray();

                using var hmac = new HMACSHA256(_providerKey);
                byte[] fullExpectedMac = hmac.ComputeHash(tokenBytes);
                byte[] expectedMac = fullExpectedMac.Take(16).ToArray();

                if (!Shield.ConstantTimeEquals(receivedMac, expectedMac, 16))
                    return null;

                string json = Encoding.UTF8.GetString(tokenBytes);
                var doc = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(json);

                var result = new Dictionary<string, object>();
                foreach (var kvp in doc)
                {
                    result[kvp.Key] = JsonElementToObject(kvp.Value);
                }
                return result;
            }
            catch
            {
                return null;
            }
        }

        private object JsonElementToObject(JsonElement element)
        {
            switch (element.ValueKind)
            {
                case JsonValueKind.String:
                    return element.GetString();
                case JsonValueKind.Number:
                    return element.GetInt64();
                case JsonValueKind.True:
                    return true;
                case JsonValueKind.False:
                    return false;
                case JsonValueKind.Null:
                    return null;
                case JsonValueKind.Array:
                    var list = new List<object>();
                    foreach (var item in element.EnumerateArray())
                        list.Add(JsonElementToObject(item));
                    return list;
                case JsonValueKind.Object:
                    var dict = new Dictionary<string, object>();
                    foreach (var prop in element.EnumerateObject())
                        dict[prop.Name] = JsonElementToObject(prop.Value);
                    return dict;
                default:
                    return element.ToString();
            }
        }

        private string GenerateNonce()
        {
            byte[] bytes = Shield.RandomBytes(8);
            return BitConverter.ToString(bytes).Replace("-", "").ToLower();
        }

        private long GetLong(Dictionary<string, object> dict, string key)
        {
            if (!dict.TryGetValue(key, out var value))
                return 0;
            return Convert.ToInt64(value);
        }

        private List<string> GetStringList(Dictionary<string, object> dict, string key)
        {
            if (!dict.TryGetValue(key, out var value))
                return new List<string>();
            if (value is List<object> objList)
                return objList.Select(o => o?.ToString() ?? "").ToList();
            if (value is List<string> strList)
                return strList;
            return new List<string>();
        }

        private Dictionary<string, object> GetDict(Dictionary<string, object> dict, string key)
        {
            if (!dict.TryGetValue(key, out var value))
                return new Dictionary<string, object>();
            if (value is Dictionary<string, object> d)
                return d;
            return new Dictionary<string, object>();
        }
    }
}
